osclib/stagingapi.py

```diff
@@ -1072,25 +1072,27 @@ def map_ring_package_to_subject(self, project, pkg):
 
         return project
 
-    def get_sub_packages(self, pkg, project=None):
+    def get_sub_packages(self, package, project=None):
         """
         Returns a list of packages that need to be linked into rings
         too. A package is actually a tuple of project and package name
         """
         ret = []
         if not project:
-            project = self.ring_packages.get(pkg)
+            project = self.ring_packages.get(package)
         if not project:
             return ret
-        url = self.makeurl(['source', project, pkg],
+        url = self.makeurl(['source', project, package],
                            {'cmd': 'showlinked'})
 
         # showlinked is a POST for rather bizzare reasons
         f = http_POST(url)
         root = ET.parse(f).getroot()
 
         for pkg in root.findall('package'):
-            ret.append((pkg.get('project'), pkg.get('name')))
+            # ensure sub-package is valid
+            if pkg.get('project') in self.rings and pkg.get('name') != package:
```

Comment:
Is `pkg.get('project') in self.rings` a problem for `adi` trying to bring in subpackages?

Reply:
Much of the problems with this code base is shooting future developers in the foot. The method name is `get_sub_packages()` and getting sub packages applies to both adi and non-adi packages. The fact that it is currently not used on adi packages doesn't mean we should intentionally make it only work on non-adi packages. If it was a load of extra work I would agree, deal with it in the future, but this is simply short-sighted and due to not properly defining the problem. The problem is validating that the links are not to external projects not that they are links to ring projects.

Most of the features I added requiring changing massive amounts of code that did similar things, rather than actually solving the problem the method implied it was solving. This is the definition of technical debt.

Reply:
The description of `get_sub_packages()` is **Returns a list of packages that need to be linked into rings too.**

I will not continue the discussion here any further. You can revert it, change it to what you've expected, or whatsoever. Just note that, the issue here is - for ring sub-package, the links are not to external projects and they are links to ring projects.

Reply:
Indeed, it just means I get to pay the technical debt like the rest of the things of this nature.